### PR TITLE
Add `Differentiable` requirements to pattern substitutions / pattern generic signature

### LINKED ISSUE #65487: Invalid type parameter in getReducedType()
Consider the following patch that switches off derivative function type cache:
```diff
diff --git a/lib/SIL/IR/SILFunctionType.cpp b/lib/SIL/IR/SILFunctionType.cpp
index d361a27f7d9..2b0b5a3aff7 100644
--- a/lib/SIL/IR/SILFunctionType.cpp
+++ b/lib/SIL/IR/SILFunctionType.cpp
@@ -883,7 +883,7 @@ CanSILFunctionType SILFunctionType::getAutoDiffDerivativeFunctionType(
   auto insertion =
       ctx.SILAutoDiffDerivativeFunctions.try_emplace(key, CanSILFunctionType());
   auto &cachedResult = insertion.first->getSecond();
-  if (!insertion.second)
+  if (0 && !insertion.second)
     return cachedResult;

   SILFunctionType *constrainedOriginalFnTy =
```

Then `AutoDiff/SILOptimizer/semantic_member_accessors_sil.swift` fails with:
```
Invalid type parameter in getReducedType()
Original type: τ_0_0.TangentVector
Simplified term: τ_0_0.[Differentiable:TangentVector]
Longest valid prefix: τ_0_0
Prefix type: τ_0_0

Requirement machine for <τ_0_0, τ_0_1>
Rewrite system: {
}
}
Property map: {
}
Conformance paths: {
}
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
```

So it seems that `getReducedType()` behavior somehow changes. I double checked and input types seems to be same on the first and subsequent invocations

### PR DESCRIPTION
**Pull request by asl**
when calculating constrained function type. Also, add requirements for differentiable results as well.

Fixes #65487

### CODE REVIEW

**Comment by asl** [comment]
I do not have good testcase besides the way described in https://github.com/apple/swift/issues/65487

However, the issue is very sound and will be easily reproduces when / if we will start supporting differentiation of coroutines (coroutines use pattern substitutions instead of directly substituting into component types).

I checked that `Differentiable` requirement is added properly to corresponding generic types (it turned out that the current code missed result types, so I added it as well while there).

**Comment by asl** [comment]
Tagging @BradLarson

**Comment by asl** [comment]
@swift-ci please test

File: lib/AST/AutoDiff.cpp [hunk 1]
```diff
@@ -308,21 +330,48 @@ GenericSignature autodiff::getConstrainedDerivativeGenericSignature(
...
+  for (unsigned resultIdx : diffResultIndices->getIndices()) {
+    // Handle formal original result.
+    if (resultIdx < originalFnTy->getNumResults()) {
+      auto resultType = originalResults[resultIdx].getInterfaceType();
+      addRequirement(resultType);
+      continue;
+    }
+    // Handle original semantic result parameters.
+    // FIXME: Constraint generic yields when we will start supporting them
+    auto resultParamIndex = resultIdx - originalFnTy->getNumResults();
```
Comment (asl) [inline lib/AST/AutoDiff.cpp:366]: Technically, checking semantic result parameters and adding corresponding requirements here is redundant as those are always wrt params. However, I added for the sake of completeness. And we will add here yields in a similar way when they will be supported

**Comment by asl** [comment]
So, for the function type mentioned in #65487 we're now generating:
```
(sil_function_type type=@callee_guaranteed @substituted <τ_0_0, τ_0_1 where τ_0_0 : Differentiable, τ_0_1 : Differentiable> (@inout Generic<τ_0_0>, @in_guaranteed τ_0_1) -> () for <T, T>
  (input=bound_generic_struct_type decl=null.(file).Generic@/Users/asl/Projects/swift-project/swift/test/AutoDiff/SILOptimizer/semantic_member_accessors_sil.swift:21:8
    (generic_type_param_type depth=0 index=0))
  (input=generic_type_param_type depth=0 index=1)
  (substitution_map generic_signature=<τ_0_0, τ_0_1 where τ_0_0 : Differentiable, τ_0_1 : Differentiable>
    (substitution τ_0_0 -> T)
    (substitution τ_0_1 -> T)
    (conformance type=τ_0_0
      (abstract_conformance protocol=Differentiable))
    (conformance type=τ_0_1
      (abstract_conformance protocol=Differentiable)))
  (substitution_map generic_signature=<nullptr>))
```

Note that w/o adding handling of result types only one requirement (for τ_0_0) was added.

**Comment by fibrechannelscsi** [comment]
I have a reproducer which may break this issue (this was compiled with the 09-24a toolchain):
```swift
import _Differentiation; public struct D<T>: Differentiable {}
extension D {@differentiable(reverse, wrt: self) mutating func m(r: @differentiable(reverse) (T?) -> T?) {}}
```

Error message without stack trace:
```
Invalid type parameter in getReducedType()
Original type: Optional<τ_0_0.TangentVector>
Simplified term: τ_0_0.[Differentiable:TangentVector]
Longest valid prefix: τ_0_0
Prefix type: τ_0_0

Requirement machine for <τ_0_0, τ_0_1>
Rewrite system: {
}
}
Property map: {
}
Conformance paths: {
}
```

If it turns out to not be related, I can open a separate issue.

**Comment by asl** [comment]
Thanks @fibrechannelscsi! It's indeed the reproducer for the issue here. I added the testcase to the PR

**Comment by asl** [comment]
@swift-ci please test

**Review by rxwei — APPROVED** [review]

**Pull request merged**